src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -73,6 +73,18 @@ class Parameter
         ros::param::set(getROSParameterPath(), new_value);
     }
 
+    /**
+     * Sets the local value of the parameter to the new value. This is primarily useful
+     * for unit testing, so that parameters can be adjusted without needing the ROS
+     * Parameter server
+     *
+     * @param new_value The new value to be set
+     */
+    void setValueLocally(T new_value)
+    {
+        value_ = new_value;
+    }
```

Comment:
Hmm, our "Parameters" are supposed to act as an abstraction between our system and other systems (like the ROS parameter server). This means that the interface should be ROS agnostic, ie, each parameter should really only have a `set(value)`, `get()` and maybe `update()` method (to pull from any other sources). If we want to be able to use Parameters without ROS (which we absolutely should, I fully support this sentiment), this seems like it might good case for having a `DynamicParameter` class with a `RosDynamicParameter` subclass? That said, this is definitely tricky and there are definitely other options, but it's also a great opportunity to make sure our code is ROS agnostic, which will make testing and working with it easier in the future.

Reply:
Yes, we could either have a parent and subclass, or an interface with 2 implementations. Both options seem reasonable. I think I will try the subclass, since then we won't have to duplicate function implementations (singe the `get()` function would be the same for example)..

One thing to note is that this means the function names will be the same, so each implementation will have a `set` function and we can't call the ROS one `setValueInROSParameterServer` like you mention above. However the interaction should be clear because the entire class will be an implementation for the ROS Parameter Server.

Reply:
This still presents a problem when trying to use Parameters with Unit Tests IMO. In this case we would be using the 'LocalParameter` implementation of a Parameter so the ROS Parameter Server is involved. But if we are testing code that uses a `ROSDynamicParameter` this won't work. We wouldn't even be able to have 2 parameters with the same name (one local, one ROS) and somehow choose between them at compile/runtime.

Reply:
Yeah, I would imagine the `set` function calling the `setValueInROSParameterServer` function if you wanted to make the implementation a bit more clear. 

You make a great point about unit tests still being difficult. Even with what you suggested above, I'm still not entirely clear on how the unit tests could easily instantiate all the parameters?

Reply:
Oh that's still a good point, the compiler will require all the Parameters specified in the header to be defined. It sounds like you're thinking more of a composition design pattern?